File: code/C++/42-Trapping Rain Water.cpp
/*
Trapping Rain Water
Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.

For example, 
Given [0,1,0,2,1,0,1,3,2,1,2,1], return 6.
我一开始是想用最蠢的办法，从左往右扫一边然后找到每个能容雨的池子然后叠加，后来编码时候发现这办法真是蠢爆肝了，蠢的我天怒人怨各种要回溯修改。

然后发现一个好办法，先找到最高的一个柱子，然后在对其左右分别处理
这样至少节约了很多回溯过程，简单两边的操作都是对称的，极易处理
*/

class Solution {
public:
    int trap(int A[], int n) {
        
        int max = 0; // 最高的柱子,将数组分为两半 
          for (int i = 0; i < n; i++)
              if (A[i] > A[max]) 
                max = i;
          
          int water = 0;
          for (int i = 0, peak = 0; i < max; i++)
              if (A[i] > peak) 
                 peak = A[i];
              else
                  water += peak - A[i];
          for (int i = n - 1, top = 0; i > max; i--)
              if (A[i] > top) 
                 top = A[i];
              else 
                 water += top - A[i];
          return water;
        
    }
};

File: code/42-Trapping Rain Water/42-Trapping Rain Water.py
"""
LeetCode 42 - Trapping Rain Water

解题步骤（自动整理）
1. 先把题目目标拆成可操作的子任务（比较/统计/构造等）
2. 选择合适的数据结构并按一次遍历或分治步骤实现核心逻辑
3. 补齐边界条件（空输入、单元素、重复元素等）并返回结果
"""
# Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it is able to trap after raining.
#
#
# The above elevation map is represented by array [0,1,0,2,1,0,1,3,2,1,2,1]. In this case, 6 units of rain water (blue section) are being trapped. Thanks Marcos for contributing this image!
#
# Example:
#
# Input: [0,1,0,2,1,0,1,3,2,1,2,1]
# Output: 6



class Solution(object):
    def trap(self, height):
        """
        :type height: List[int]
        :rtype: int
        """
        maxPillar = height.index(max(height))
        res, peakLeft, peakRight = 0, 0, len(height) - 1
        for i in range(maxPillar):
            if height[i] > height[peakLeft]:
                peakLeft = i
            else:
                res += height[peakLeft] - height[i]
        for i in range(len(height) - 1, maxPillar, -1):
            if height[i] > height[peakRight]:
                peakRight = i
            else:
                res += height[peakRight] - height[i]
        return res

if __name__ == '__main__':
    s = Solution()
    print(s.trap([0,1,0,2,1,0,1,3,2,1,2,1]))
